rust-runtime/aws-smithy-runtime/src/client/orchestrator/auth.rs

```diff
@@ -190,6 +183,86 @@ pub(super) async fn orchestrate_auth(
 }
 
+pub(super) fn sign_request(
+    scheme_id: AuthSchemeId,
+    identity: &Identity,
+    ctx: &mut InterceptorContext,
+    runtime_components: &RuntimeComponents,
+    cfg: &ConfigBag,
+) -> Result<(), BoxError> {
+    trace!("signing request");
+    let request = ctx.request_mut().expect("set during serialization");
+    let endpoint = cfg
+        .load::<Endpoint>()
+        .expect("endpoint added to config bag by endpoint orchestrator");
+    let auth_scheme = runtime_components
+        .auth_scheme(scheme_id)
+        .ok_or("should be configured")?;
+    let signer = auth_scheme.signer();
+    let auth_scheme_endpoint_config = extract_endpoint_auth_scheme_config(endpoint, scheme_id)?;
+    trace!(
+        signer = ?signer,
+        "signing implementation"
+    );
+    signer.sign_http_request(
+        request,
+        identity,
+        auth_scheme_endpoint_config,
+        runtime_components,
+        cfg,
+    )?;
+    Ok(())
+}
+
+// Marker indicating that endpoint resolution for determining the auth scheme option
+// has already been performed by a `AuthSchemeOptionResolver`, such as `EndpointBasedAuthSchemeOptionResolver`.
+//
+// When this marker is present in the config bag (the default for SDKs moving forward),
+// `resolve_identity` will skip endpoint resolution.
+#[doc(hidden)]
+#[derive(Clone, Debug)]
+pub struct AuthSchemeOptionResolverHandlesEndpointResolution;
+
+impl Storable for AuthSchemeOptionResolverHandlesEndpointResolution {
+    type Storer = StoreReplace<Self>;
+}
+
+// Checks whether an `endpoint` resolved by `SharedEndpointResolver` in `runtime_components`,
+// using `EndpointResolverParams` from `cfg`, contains `scheme_id` in its `authSchemes` property.
+//
+// Returns `Ok` if the condition is met, with the `Ok` variant also containing the actual `endpoint` used for verification.
+//
+// This function exists for backward compatibility. SDKs generated with the `AuthSchemeOptionResolverHandlesEndpointResolution`
+// do not, by default, require an endpoint outside of `SharedAuthSchemeOptionResolver`,
+// so the function short-circuits and returns `Ok(None)`.
+async fn endpoint_auth_scheme_matches_configured_scheme_id(
+    runtime_components: &RuntimeComponents,
+    cfg: &ConfigBag,
+    scheme_id: AuthSchemeId,
+) -> Result<Option<Endpoint>, AuthOrchestrationError> {
+    if cfg
+        .load::<AuthSchemeOptionResolverHandlesEndpointResolution>()
+        .is_some()
+    {
+        // Short-circuit if auth scheme option resolver internally handles endpoint resolution
+        return Ok(None);
+    }
+
+    let params = cfg
+        .load::<EndpointResolverParams>()
+        .expect("endpoint resolver params must be set");
+
+    let endpoint = runtime_components
+        .endpoint_resolver()
+        .resolve_endpoint(params)
+        .await
+        .map_err(AuthOrchestrationError::FailedToResolveEndpoint)?;
+
+    let _ = extract_endpoint_auth_scheme_config(&endpoint, scheme_id)?;
```

Comment:
Why do we have to call this here and discard the result?

Reply:
Ah, good catch. Maybe I should either add a comment or refactor so the intent is clear. What we actually want to verify here is [this snippet](https://github.com/smithy-lang/smithy-rs/blob/fd59551fb4de790cf427cb8ad3bcd64dcd5a8f75/rust-runtime/aws-smithy-runtime/src/client/orchestrator/auth.rs#L285-L294) within `extract_endpoint_auth_scheme_config`. At its core, we want this predicate:
```
auth_schemes.iter().any(|doc| {...})
```
but I didn't bother refactoring to extract that part out of `extract_endpoint_auth_scheme_config` since line 261 (or after 251 to be more precise) only takes care of the legacy logic.

Let me know, I can go either way (adding a comment or refactor the predicate).

Reply:
Ah, good catch. Maybe I should either add a comment or refactor so the intent is clear. What we actually want to verify here is [this snippet](https://github.com/smithy-lang/smithy-rs/blob/fd59551fb4de790cf427cb8ad3bcd64dcd5a8f75/rust-runtime/aws-smithy-runtime/src/client/orchestrator/auth.rs#L285-L294) within `extract_endpoint_auth_scheme_config`. At its core, we want this predicate:
```
auth_schemes.iter().any(|doc| {...})
```
but I didn't bother refactoring to extract that part out of `extract_endpoint_auth_scheme_config` since line 261 above (or lines after 251 to be more precise) only takes care of the legacy logic.

Let me know, I can go either way (adding a comment or refactor the predicate).

Reply:
Added comment in 41e31d0